docs/rfcs/0001-image-watcher.md

```diff
@@ -0,0 +1,82 @@
+- Start Date: 2020-11-18
+- Target Version: 1.0
+
+# Summary
+This RFC proposes adding a new feature to trigger Deployment when an image tag stored at a container registry is updated.
+Similar to Flux's [Image Scan](https://docs.fluxcd.io/en/1.19.0/faq/#how-often-does-flux-check-for-new-images).
+
+# Motivation
+The canonical deployment flow with PipeCD is:
+
+1. CI server pushes the generated image to the container registry after app-repo updated.
+2. User updates the config-repo manually.
+
+It is the User's responsibility to automate these 1 and 2 step to be done in a series of actions, while it is a bit quite painful.
+Therefore, we would like to solve this problem by letting Piped monitor the container registry.
+
+# Detailed design
+Piped compares the image tags defined in git with the latest tags stored in the container registry, and then pushes them to git if there are any deviations.
+
+### Usage
+Note: Where, we call the client that accesses the container registry to `Image Provider`.
+
+First of all, we need to register the credentials for accessing the image provider with Piped.
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: Piped
+spec:
+  imageProviders:
+    - name: my-dockerhub
+      type: DOCKERHUB
+      # Indicates how many minutes Piped polls every.
+      interval: 5m
+      config:
+        username: foo
+        passwordFile: /etc/piped-secret/dockerhub-pass
+```
+
+We define the information about image tag underneath `.pipe/` directory:
```

Comment:
What happens when multiple `piped`s handling the same configuration repository?

Reply:
I feel like no problem even in that case. Image Watcher does only sync the tag in Git with one in Container Registry.  If multiple Piped push to the Git repo, only one push will be successful.

And I'd say the case of using multiple Piped for the single repo is an edge case, but what do you think?

Reply:
Depends on the way Piped commits the changes. If it directly pushes the change to the specified branch, it is ok for handling the pushing error. If it pushes the change via PR, the conflicted pull requests will remain.

Alternatively, we can update the piped configuration to watch only the specified files in the `.piped` directory.

Reply:
I guess say we can consider it is the user's responsibility to ensure that one image is watched by one Piped, but how about you?

Reply:
Yes, I totally agree that it should be done by the user, yes their responsibility.
Just want to make sure we can provide a way for them to get it right.

For the use case at our company, the AB... team is going to store all configuration in a single repository, and I guess they will have at least 4 pipeds handling that repo. All of them run separately and loading the same `.pipe` directory.

Reply:
Initially, I thought the configuration assosiated to application should be defined in a repo, to clearly separate the roles of the user and Piped Operator.

However, Image Watcher is obviously a feature that piped is responsible for, I'm starting to feel like all the configurations (including the filepath in the target repo) should be written in piped config:
```yaml
apiVersion: pipecd.dev/v1beta1
kind: Piped
spec:
  ...
  repositories:
    - repoId: foo
      remote: git@github.com:nakabonne/foo.git
      branch: master
  imageProviders:
    - name: my-dockerhub
      type: DOCKERHUB
      interval: 5m
      config:
        username: user
        passwordFile: /etc/piped-secret/dockerhub-pass
  imageWatcher:
    targets:
    - image: nakabonne/podinfo
      path:
        repoId: foo
        filePath: podinfo/deployment.yaml
        field: spec.template.spec.containers[0].image
      provider: my-dockerhub
```

This also solves the problem of running a large number of pulls when there are a bunch of repositories, which was mentioned in the unresolved question section.

While a little out of the concept, this significantly reduces the chance of unexpected conflicts, even in cases where multiple Piped handle a single repository.

Reply:
> This also solves the problem of running a large number of pulls when there are a bunch of repositories, which was mentioned in the unresolved question section.

I really don't worry too much about this problem. Because:
- we are having a cache for our git client
- user can break into multi pipeds when they reach the limit


>However, Image Watcher is obviously a feature that piped is responsible for, I'm starting to feel like all the configurations (including the filepath in the target repo) should be written in piped config:

I agree that this feature should not be a part of the application.
But putting them into the Piped configuration is not so good.
Because piped is maintained by the operators and they don't know where are the files should be updated, what applications should be enabled...

So I think we should keep putting those configurations under the `.pipe` directory and by default piped will load all configuration files inside that directory. But we also provide a configuration field in Piped configuration to allow the operators to specify what file should be loaded, what file should be ignored.

Reply:
And putting them into the Piped configuration means the piped has to be restarted/deployed every time when enabling that feature for a new application.

Reply:
Right, totally agree with you. That is, something like this?:

```yaml
apiVersion: pipecd.dev/v1beta1
kind: Piped
spec:
  ...
  repositories:
    - repoId: foo
      remote: git@github.com:nakabonne/foo.git
      branch: master
  imageProviders:
    - name: my-dockerhub
      type: DOCKERHUB
      interval: 5m
      config:
        username: user
        passwordFile: /etc/piped-secret/dockerhub-pass
# Optional
  imageWatcher:
    ignoreFiles:
      - repoId: foo
        filePath: podinfo/deployment.yaml
# Or only one of them is available
#   loadFiles:
#     - repoId: foo
#       filePath: podinfo/deployment.yaml
```

Reply:
I think something like this

``` yaml
apiVersion: pipecd.dev/v1beta1
kind: Piped
spec:
  ...
  repositories:
    - repoId: foo
      remote: git@github.com:nakabonne/foo.git
      branch: master

  imageProviders:
    - name: my-dockerhub
      type: DOCKERHUB
      interval: 5m
      config:
        username: user
        passwordFile: /etc/piped-secret/dockerhub-pass

# Optional
  imageWatcher:
        - repoId: foo
           excludes:
           -  .pipe/imagewatcher.yaml
```

Reply:
Looks good, that's more conventional.

Reply:
Ah, you now found you specify `.pipe/imagewatcher.yaml` as an excluded file. Does it mean to ignore/load the entire repository instead of specifying a specific application?

Reply:
It means the including/excluding can be done at the image-watcher file level, not the app/manifest-path level.
For a big repository, if they want more control on multiple pipeds they should split into multiple image-watcher files inside the `.pipe` directory.

Reply:
Let me tell you the case I consider.

Let's say that multiple environment settings are in the same repository:
```
repo-root
├── dev/
├── prod/
```

Each piped should only monitor images for their environment.

dev piped should be:
```yaml
apiVersion: pipecd.dev/v1beta1
kind: Piped
spec:
  ...
  repositories:
    - repoId: foo
      remote: git@github.com:nakabonne/foo.git
      branch: master

  imageProviders:
    - name: my-dockerhub
      type: DOCKERHUB
      interval: 5m
      config:
        username: user
        passwordFile: /etc/piped-secret/dockerhub-pass

# Optional
  imageWatcher:
    - repoId: foo
      includes:
       - dev/
```

prod piped config should be:
```yaml
``yaml
apiVersion: pipecd.dev/v1beta1
kind: Piped
spec:
  ...
  repositories:
    - repoId: foo
      remote: git@github.com:nakabonne/foo.git
      branch: master

  imageProviders:
    - name: my-dockerhub
      type: DOCKERHUB
      interval: 5m
      config:
        username: user
        passwordFile: /etc/piped-secret/dockerhub-pass

# Optional
  imageWatcher:
    - repoId: foo
      includes:
       - prod/
```

Reply:
>For a big repository, if they want more control on multiple pipeds they should split into multiple image-watcher files inside the .pipe directory.

That makes sense!